Mathlib/Algebra/BigOperators/Group/Finset/Defs.lean

```diff
@@ -231,47 +231,91 @@ end deprecated
+
+/-- Delaborates a finset indexing a big operator. In case it is a `Finset.filter`, `i` is used for
+the binder name. -/
+private def delabFinsetArg (i : Ident) : DelabM FinsetResult := do
+  let s ← getExpr
+  if s.isAppOfArity ``Finset.univ 2 then
+    return .univ
+  else if s.isAppOfArity ``Finset.filter 4 then
+    let #[_, _, _, t] := s.getAppArgs | failure
+    let p ←
+      withNaryArg 1 do
+        if (← getExpr).isLambda then
+          withBindingBody i.getId delab
+        else
+          let p ← delab
+          return (← `($p $i))
+    if t.isAppOfArity ``Finset.univ 2 then
+      return .filterUniv p
+    else
+      let ss ← withNaryArg 3 delab
+      return .filter ss p
+  else
+    let ss ← delab
+    return .finset ss
+
 /-- Delaborator for `Finset.prod`. The `pp.piBinderTypes` option controls whether
 to show the domain type when the product is over `Finset.univ`. -/
 @[app_delab Finset.prod] def delabFinsetProd : Delab :=
   whenPPOption getPPNotation <| withOverApp 5 <| do
-  let #[_, _, _, s, f] := (← getExpr).getAppArgs | failure
-  guard <| f.isLambda
+  let #[_, _, _, _, f] := (← getExpr).getAppArgs | failure
+  guard f.isLambda
   let ppDomain ← getPPOption getPPPiBinderTypes
   let (i, body) ← withAppArg <| withBindingBodyUnusedName fun i => do
-    return (i, ← delab)
-  if s.isAppOfArity ``Finset.univ 2 then
+    return (⟨i⟩, ← delab)
+  let res ← withNaryArg 3 <| delabFinsetArg i
+  match res with
+  | .finset ss => `(∏ $i:ident ∈ $ss, $body)
+  | .univ =>
     let binder ←
       if ppDomain then
         let ty ← withNaryArg 0 delab
-        `(bigOpBinder| $(.mk i):ident : $ty)
+        `(bigOpBinder| $i:ident : $ty)
       else
-        `(bigOpBinder| $(.mk i):ident)
+        `(bigOpBinder| $i:ident)
     `(∏ $binder:bigOpBinder, $body)
-  else
-    let ss ← withNaryArg 3 <| delab
-    `(∏ $(.mk i):ident ∈ $ss, $body)
+  | .filter ss p =>
+    `(∏ $i:ident ∈ $ss with $p, $body)
+  | .filterUniv p =>
+    `(∏ $i:ident with $p, $body)
 
 /-- Delaborator for `Finset.sum`. The `pp.piBinderTypes` option controls whether
 to show the domain type when the sum is over `Finset.univ`. -/
 @[app_delab Finset.sum] def delabFinsetSum : Delab :=
   whenPPOption getPPNotation <| withOverApp 5 <| do
-  let #[_, _, _, s, f] := (← getExpr).getAppArgs | failure
+  let #[_, _, _, _, f] := (← getExpr).getAppArgs | failure
   guard <| f.isLambda
   let ppDomain ← getPPOption getPPPiBinderTypes
   let (i, body) ← withAppArg <| withBindingBodyUnusedName fun i => do
-    return (i, ← delab)
-  if s.isAppOfArity ``Finset.univ 2 then
+    return ((⟨i⟩ : Ident), ← delab)
+  let res ← withNaryArg 3 <| delabFinsetArg i
+  match res with
+  | .finset ss => `(∑ $i:ident ∈ $ss, $body)
+  | .univ =>
     let binder ←
-      if ppDomain then
-        let ty ← withNaryArg 0 delab
-        `(bigOpBinder| $(.mk i):ident : $ty)
-      else
-        `(bigOpBinder| $(.mk i):ident)
+    if ppDomain then
+      let ty ← withNaryArg 0 delab
```

Comment:
Let's test for the behaviour with piBinderTypes on/off. I also think the option should control what happens in the filterUniv case, in the same way as it does here (and that should be tested too)